documentation/design_proposal/002_grafana_v1beta1.md

```diff
@@ -0,0 +1,191 @@
+In openshift a good default is TLS enabled by default. While in ingress it's not as easy.
+
+So what needs to change?
+
+The default setting for routes should be [edge](https://docs.openshift.com/container-platform/4.9/networking/routes/secured-routes.html) `TLS enabled=true` but enable the possability to run in http mode only.
+
+The default ingress should keep on being TLS enabled=false
+
+#### Default resource requests
+
+To miniamize the risk of our operator and the grafana deployment to be evicted during high load in your clusters we should provide basic resource reuqests for both the operator and the grafana deployment/init container.
+
+## Alternatives
+
+Instead of having a custom object for grafanaContainer we could provide the user with the entire deployment object with some tweaks and have them configure everything from inside the deployment.
+
+This would make it harder to use but probably allot easier from a development point of view since we would have to do less advanced merges between configs.
+
+```.go
+type GrafanaSpec struct {
+ Config                     GrafanaConfig            `json:"config"`
+ DashboardLabelSelector     []*metav1.LabelSelector  `json:"dashboardLabelSelector,omitempty"`
+ DeploymentOverrides        *typeoverride.Deployment `json:"deploymentOverrides,omitempty"`
+ Ingress                    *GrafanaIngress          `json:"ingress,omitempty"`
+ Service                    *GrafanaService          `json:"service,omitempty"`
+ ServiceAccount             *GrafanaServiceAccount   `json:"serviceAccount,omitempty"`
+ Client                     *GrafanaClient           `json:"client,omitempty"`
+ DashboardNamespaceSelector *metav1.LabelSelector    `json:"dashboardNamespaceSelector,omitempty"`
+ Jsonnet                    *JsonnetConfig           `json:"jsonnet,omitempty"`
+}
+
+// Deployment is a subset of [Deployment in k8s.io/api/apps/v1](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.19/#deployment-v1-apps), with [DeploymentSpec replaced by the local variant](#deployment-spec).
+type Deployment struct {
+ ObjectMeta `json:"metadata,omitempty"`
+ // The desired behavior of [this deployment](#deploymentspec).
+ Spec DeploymentSpec `json:"spec,omitempty"`
+}
+```
+
+[DeploymentSpec](https://github.com/banzaicloud/operator-tools/blob/2189d8efc3856efd4a7c7fbb28b7cba9a977d0bd/pkg/typeoverride/override.go#L137-L176)
+
+It will be harder to do basic things like setting baseImages.
+
+Instead of setting `grafana.spec.baseImage` you would have to do `grafana.spec.deploymentOverrides.DeploymentSpec.templates.containers.grafana.image`.
+
+## Work Plan
+
+Provide this PR with `grafana_types.go` containing all the type definitions.
+Create a PR that implements this solution.
+
+## Open questions
+
+- Should we do a similar solution for the init pluigin container?
```

Comment:
I would say _no_, the init container is very, very basic, and we seldom get any issues with it